File: web/ray-tracer-ts/scene.ts
import { Vec3 } from "./vec3";
import { Sphere, HittableList, Lambertian, Metal } from "./objects";
import { renderImage } from "./render";

export const width = 400;
export const height = 300;

export function createScene(): HittableList {
  // Define materials
  const grayMaterial = new Lambertian(new Vec3(0.5, 0.5, 0.5));
  const redMaterial = new Lambertian(new Vec3(0.8, 0.2, 0.2));
  const greenMetal = new Metal(new Vec3(0.6, 0.8, 0.2), 0.0);
  const blueMetal = new Metal(new Vec3(0.2, 0.5, 0.8), 0.3);

  const objects = [
    // Three spheres in front - mix of materials
    new Sphere(new Vec3(0, 0, -1), 0.5, redMaterial),
    new Sphere(new Vec3(-1, 0, -1), 0.5, greenMetal),
    new Sphere(new Vec3(1, 0, -1), 0.5, blueMetal),
    // Ground sphere
    new Sphere(new Vec3(0, -100.5, -1), 100, grayMaterial),
  ];

  // Flatten nested structure (F# had nested HittableList)
  return new HittableList(objects);
}

export function render(): Float64Array {
  const world = createScene();
  return renderImage(world, {
    width,
    height,
    origin: new Vec3(0, 0, 1),
    focalLength: 1.9,
    samplesPerPixel: 100,
    maxRayBounces: 20,
  });
}

File: src/raytracer_raw/Scene.scala
package raytracer_raw

// Direct port of TypeScript scene.ts
// Mirrors web/ray-tracer-ts/scene.ts EXACTLY

object Scene {

  def createScene(): Hittable = {
    // Define materials - exact match to TypeScript
    val grayMaterial = Lambertian(Vec3(0.5, 0.5, 0.5))
    val redMaterial = Lambertian(Vec3(0.8, 0.2, 0.2))
    val greenMetal = Metal(Vec3(0.6, 0.8, 0.2), fuzz = 0.0)
    val blueMetal = Metal(Vec3(0.2, 0.5, 0.8), fuzz = 0.3)

    val objects = Seq(
      // Three spheres in front - mix of materials
      Sphere(Vec3(0.0, 0.0, -1.0), 0.5, redMaterial),
      Sphere(Vec3(-1.0, 0.0, -1.0), 0.5, greenMetal),
      Sphere(Vec3(1.0, 0.0, -1.0), 0.5, blueMetal),
      // Ground sphere
      Sphere(Vec3(0.0, -100.5, -1.0), 100.0, grayMaterial)
    )

    HittableList(objects)
  }
}
